src/Servers/ShellServer.cs

```diff
@@ -140,7 +143,7 @@ private void EventLoop(NetMQSocket socket)
 
                     // Get a service that can handle the message type and
                     // dispatch.
-                    router.Handle(nextMessage);
+                    Task.Run(() => router.HandleAsync(nextMessage));
```

Comment:
Is it really ok to run all messages asynchronously?
Maybe some messages need to be processed in order by their corresponding router/handler.
So it might be better to have the router to deal with that, maybe just having a method to accept messages (StartHandleMessage), and the router decides whether to create new tasks for and run them async or not.
The same logic could be used for the handler, so the router does not make that decision, but the handler does.
This way we could give more control to the handler implementation on how to process messages (concurrently or not).
From the ShellServer and Router perspective, they are just starting things and they don't wait for the message to be processed, only for the message to be accepted.

Reply:
The trouble is that even when a message needs to be handled synchronously with respect to other messages of the same kind, messages of other message types still need to be processed asynchronously. We've not done that correctly in previous versions, but as we increase the amount of client communication, that becomes more of an issue.

Reply:
I've modified to using a queue, and to having handlers return either `null` to indicate no asynchronous completion is required, or a task that indicates that the remainder of handling can be performed asynchronously. This allows for ensuring messages are always properly ordered, but that processing can be nonblocking for long-running handlers (e.g.: execution).